csrc/opaque_type.h

```diff
@@ -0,0 +1,71 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <any>
+#include <functional>
+#include <ostream>
+
+namespace nvfuser {
+
+class Opaque;
+
+template <typename T>
+struct OpaqueEquals {
+  bool operator()(const Opaque& a, const Opaque& b) const;
+};
+
+class Opaque {
+  std::any value_;
+  std::function<bool(const Opaque&, const Opaque&)> equals_;
+
+ public:
+  template <typename T>
+  explicit Opaque(T value)
+      : value_(std::move(value)), equals_(OpaqueEquals<T>{}) {}
```

Comment:
So currently `equals_` cannot be overridden, but if there is a need to in the future then the appropriate ctor arg or accessor methods can be added?

Reply:
I think the purpose for `equals_` here is not to allow users to customize equality check. If the user does want to do so, the user should define their own class with custom `operator==`, and store that class with `Opaque`. The only purpose for the `equals_` here is to solve the following issue:
```C++
std::any a = 3, b = 3;
a == b; // compilation error
```

Reply:
So the `equals_` determines how to dispatch `==` in the constructor, since we can't template `operator==` over `T` with runtime type. Makes sense to me.

Reply:
A bit confused here. If we don't want to allow the user to customize the equality operator, why does the equality operator need to be defined as a separate class?

Reply:
> So the equals_ determines how to dispatch == in the constructor, since we can't template operator== over T with runtime type. Makes sense to me.

This makes to me. Thanks @jacobhinkle.

Reply:
> does the equality operator need to be defined as a separate class?

It doesn't needs to. It is mostly just a matter of coding style. We can just use lambda here. Shall I change it?

Reply:
> It doesn't needs to. It is mostly just a matter of coding style. We can just use lambda here. Shall I change it?

No, don't bother. I was just wondering why we can't just do:

```
bool operator==(const Opaque& other) const {
    if (this == &other) {
      return true;
    }
    if (value_.type() != other.value_.type()) {
      return false;
    }
    return as<T>() == other->as<T>();
  }
```

But I realized we don't have `T` defined.